pkg/network/protocols/postgres/telemetry.go

```diff
@@ -97,3 +97,135 @@ func (t *Telemetry) Count(tx *ebpf.EbpfEvent, eventWrapper *EventWrapper) {
+	return &tlsAwareCounter{
+		counterPlain: metricGroup.NewCounter(metricName, append(tags, "encrypted:false")...),
+		counterTLS:   metricGroup.NewCounter(metricName, append(tags, "encrypted:true")...),
+	}
+}
+
+// Add adds the given delta to the counter based on the encryption.
+func (c *tlsAwareCounter) add(delta int64, isTLS bool) {
+	if isTLS {
+		c.counterTLS.Add(delta)
+		return
+	}
+	c.counterPlain.Add(delta)
+}
+
+// Get returns the counter value based on the encryption.
+func (c *tlsAwareCounter) get(isTLS bool) int64 {
+	if isTLS {
+		return c.counterTLS.Get()
+	}
+	return c.counterPlain.Get()
+}
+
+// PgKernelTelemetry	provides empirical kernel statistics about the number of messages in each TCP packet
+type PgKernelTelemetry struct {
+	// metricGroup is used here mostly for building the log message below
+	metricGroup *libtelemetry.MetricGroup
+	// pgMessagesCountBuckets		count Postgres messages and divide the count into buckets
+	pgMessagesCountBuckets [pgebpf.KerMsgCountNumBuckets]*tlsAwareCounter
+	// pgKernelMsgCountPlainPrev	save the last counters observed from the kernel, plain messages
+	pgKernelMsgCountPlainPrev pgebpf.PostgresKernelMsgCount
+	// pgKernelMsgCountPrev			save the last counters observed from the kernel, TLS messages
+	pgKernelMsgCountTlsPrev pgebpf.PostgresKernelMsgCount
+}
+
+// newPgKernelTelemetry this is the Postgres message counter store.
+func newPgKernelTelemetry() *PgKernelTelemetry {
+	metricGroup := libtelemetry.NewMetricGroup("usm.postgres", libtelemetry.OptStatsd)
+	pgKernelTel := &PgKernelTelemetry{
+		metricGroup: metricGroup,
+	}
+	pgKernelTel.createMessagesCountBuckets(metricGroup)
+
+	return pgKernelTel
+}
+
+func (t *PgKernelTelemetry) createMessagesCountBuckets(metricGroup *libtelemetry.MetricGroup) {
+	for bucketIndex := range t.pgMessagesCountBuckets {
+		t.pgMessagesCountBuckets[bucketIndex] = newTLSAwareCounter(metricGroup,
+			"messages_count_bucket_"+(strconv.Itoa(bucketIndex+1)))
+	}
+}
+
+// update the Postgres message counter store with new counters from the kernel.
+// return if nothing to add, in this case also metric group summary ignores unmodified metrics.
+func (t *PgKernelTelemetry) update(kernMsgCounts *pgebpf.PostgresKernelMsgCount, isTLS bool) {
+	if kernMsgCounts == nil {
+		return
+	}
+	if isEmptyCounts(kernMsgCounts) {
+		log.Debugf("postgres kernel telemetry empty.")
+		return
+	}
+	delta := t.getDiff(kernMsgCounts, isTLS)
+	if isEmptyCounts(delta) {
+		log.Debugf("postgres kernel telemetry, no difference with previous counters.")
+		return
+	}
+	for bucketIndex := range t.pgMessagesCountBuckets {
+		t.pgMessagesCountBuckets[bucketIndex].add(int64(delta.Messages_count_buckets[bucketIndex]), isTLS)
+	}
+	if isTLS {
+		t.pgKernelMsgCountTlsPrev = *kernMsgCounts
+	} else {
+		t.pgKernelMsgCountPlainPrev = *kernMsgCounts
+	}
+	s := t.metricGroup.Summary()
+	log.Debugf("postgres kernel telemetry, isTLS=%t, summary: %s", isTLS, s)
+}
```

Comment:
Unifying all different comments -
I'm not sure what's the benefit of that implementation

The user-mode telemetry should be a mirror of the ebpf's, so we need to pull the current ebpf-state, and store it in user mode.

What's happening today (also relevant to http2), is that we have `currentTelemetry`, `newTelemetry` and we do
```go
delta := newTelemetry - currentTelemetry
currentTelemetry = currentTelemetry + delta
```

That's redundant, and equals to `cuurentTelemetry = newTelemetry`

Reply:
The idea was not to proceed further and exit at line 176 is there is no difference in counters obtained from kernel. I can remove it if you think it is not right.

Reply:
I do think we should simplify the function, as the current state of the PR is that we added a lot of code that can be deleted

Reply:
I remove usage of "delta" in recent commit